### PR TITLE
VGG19 on Imagenet dataset.

### PR DESCRIPTION
**Pull request by MuLx10**
Thus PR implements a model for tiny imagnet classification using VGG19.
I have started the training process. I'll update the tutorial with results.

### CODE REVIEW

**Comment by MuLx10** [comment]
> Looks like the indentation is off for some files.

Corrected.

File: VGG19/VGG19.hpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+/**
+ * VGG-19 is a convolutional neural network that is trained on more than a
+ * million images from the ImageNet database. The network is 19 layers
+ * deep and can classify images into 1000 object categories. Details about
+ * the network architecture can be found in the following arXiv paper:
+ *
+ * Very Deep Convolutional Networks for Large-Scale Image Recognition
+ * K. Simonyan, A. Zisserman
+ * arXiv:1409.1556
+ * https://arxiv.org/abs/1409.1556
```
Comment (zoq) [inline VGG19/VGG19.hpp]: Can you use the same cite style as used here:

https://github.com/mlpack/mlpack/blob/1174e84eebfb0181913f7ef57fc16a7f79bafd67/src/mlpack/methods/ann/layer/elu.hpp#L52-L61

File: VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+void Dataloader::LoadTrainData(const std::string folderPath, bool shuffle)
+{
+  size_t count = 1;
+
+  fs::path p(folderPath);
+  fs::directory_iterator end_itr;
+  
+  // cycle through the directory
```
Comment (zoq) [inline VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp]: Missing stop at the end.

File: VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+      }
+      count++;
+    }
+  }
+  if(shuffle)
+  {
+    auto seed = unsigned ( std::time(0) );
+
+    std::srand(seed);
+    std::random_shuffle(trainX.begin(), trainX.end());
```
Comment (zoq) [inline VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp]: Not sure this is needed, as we shuffle the data during training.

File: VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    std::srand(seed);
+    std::random_shuffle(trainY.begin(), trainY.end());
+  }
+  if (numClasses == 0)
+    numClasses = count-1;
+}
+
+void Dataloader::LoadValData(const std::string folderPath)
+{
+
```
Comment (zoq) [inline VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp]: We can remove the empty line here.

File: VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  std::ifstream dataFile(folderPath+"/val_annotations.txt");
+  while (!dataFile.eof())
+  {
+    std::string str;
+    std::getline( dataFile, str);
+    std::stringstream buffer(str);
+    std::string temp;
+    std::vector<std::string> values;
+   
+    while (getline( buffer, temp, '\t'))
```
Comment (zoq) [inline VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp]: We can remove the extra whitespace before ``buffer``.

File: VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+void Dataloader::LoadValData(const std::string folderPath)
+{
+
+  std::ifstream dataFile(folderPath+"/val_annotations.txt");
+  while (!dataFile.eof())
+  {
+    std::string str;
+    std::getline( dataFile, str);
```
Comment (zoq) [inline VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp]: See comment below.

File: VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+template<typename dataType, typename labelType>
+void Dataloader::LoadImageData(arma::Mat<dataType>& X,
+                   arma::Mat<labelType>& y,
+                   bool train,
+                   size_t limit,
+                   size_t offset)
+{
+  arma::Mat<unsigned char> colImg;
+    std::vector<size_t> tempY;
```
Comment (zoq) [inline VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp]: We can remove the extra whitespace here.

File: VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    if (values.size() == 0)
+      break;
+    valX.push_back(folderPath+"/images/"+values[0]);
+    valY.push_back(targetName[values[1]]);
+  }
+}
+
+template<typename dataType, typename labelType>
+void Dataloader::LoadImageData(arma::Mat<dataType>& X,
+                   arma::Mat<labelType>& y,
```
Comment (zoq) [inline VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp]: The parameter should align with the first one.

File: VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+                   bool train,
+                   size_t limit,
+                   size_t offset)
+{
+  arma::Mat<unsigned char> colImg;
+    std::vector<size_t> tempY;
+  data::ImageInfo info(64, 64, 3);
+
+  std::vector<std::string> dataset;
+    std::vector<size_t> labels;
```
Comment (zoq) [inline VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp]: Looks like the following is off by some spaces.

File: VGG19/vgg19_tiny_imagenet/src/model.cpp [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+using namespace std;
+using namespace ens;
+
+// Convenience typedefs
+typedef FFN<NegativeLogLikelihood<>, XavierInitialization> VGGModel;
+
+// Calculates Log Likelihood Loss over batches.
+template<typename NetworkType = FFN<NegativeLogLikelihood<>, XavierInitialization>,
+         typename DataType = arma::mat>
+double NLLLoss(NetworkType& model, DataType& testX, DataType& testY, size_t batchSize)
```
Comment (zoq) [inline VGG19/vgg19_tiny_imagenet/src/model.cpp]: Looks like we could use ``const`` here.

File: VGG19/vgg19_tiny_imagenet/src/model.cpp [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+    cout << "Epoch " << i << endl;
+    // Don't reset optimizers parameters between cycles.
+    optimizer.ResetPolicy() = false;
+
+    std::cout << "Loss after cycle " << i << " -> " 
+        << NLLLoss<VGGModel>(model, testX, testY, 50) << std::endl;
+  }
+
+  std::cout << "Time taken to train -> " << float(clock() - begin_time) /
```
Comment (zoq) [inline VGG19/vgg19_tiny_imagenet/src/model.cpp]: You could use armadillo's ``wall_clock`` with ``tic, toc``.

File: VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp (hunk 3 above)
Comment (MuLx10) [reply VGG19/vgg19_tiny_imagenet/src/dataloader_impl.hpp]: Needed if we train on first few (say 1000) samples.

**Comment by MuLx10** [comment]
@zoq The build is failing because of boost filesystem.

**Comment by zoq** [comment]
I see, can you upgrade the travis and CMake config so that it links against boost filesystem?

**Comment by MuLx10** [comment]
> I see, can you upgrade the travis and CMake config so that it links against boost filesystem?
>
Sure.

> Also, before we merge this in, we should probably wait for the results. What do you think?

Yes, it would be better to merge after I include the results in the tutorials.

**Pull request closed without merging**